In [48]:
from pathlib import Path as p
from IPython.display import display

import pandas as pd

In [49]:
def list_files(dir):
    for item in sorted(dir.iterdir()):
        item_type = 'folder' if item.is_dir() else 'file'
        print(f"- [{item_type}] {item.name}")

curr_dir = p.cwd()

print(f"Current working directory: {curr_dir}")
print("\n Files and Folders found: ")

list_files(curr_dir)

Current working directory: /home/kingcole/Documents/uniAssignments/year_3_sem_1/ai_week_5

 Files and Folders found: 
- [folder] .git
- [folder] .idea
- [folder] .venv
- [file] Assignment 4 - AirQo PM2.5 Prediction.pdf
- [file] Sensor Data Jan - July 2026.csv
- [file] requirements.txt
- [file] train.ipynb


In [50]:
data_file = p('Sensor Data Jan - July 2026.csv')
data = pd.read_csv(data_file)

In [51]:
print(f"Rows: {data.shape[0]}")
print(f"Columns: {data.shape[1]}")
print(f"Column Names: {data.columns.tolist()}")
print()

Rows: 14600
Columns: 12
Column Names: ['site_name', 'datetime', 'frequency', 'network', 'latitude', 'longitude', 'temperature', 'humidity', 'pm2_5_calibrated_value', 'pm10_calibrated_value', 'site_id', 'device_name']



In [52]:
print("Data Types; ")
print(data.dtypes)

Data Types; 
site_name                     str
datetime                      str
frequency                     str
network                       str
latitude                  float64
longitude                 float64
temperature               float64
humidity                  float64
pm2_5_calibrated_value    float64
pm10_calibrated_value     float64
site_id                       str
device_name                   str
dtype: object


In [53]:
measurement_columns = [
    'latitude',
    'longitude',
    'temperature',
    'humidity',
    'pm2_5_calibrated_value',
    'pm10_calibrated_value'
]

In [54]:
missing_values = pd.DataFrame({
    'null_values' : data[measurement_columns].isna().sum(),
    'zero_values' : (data[measurement_columns]==0).sum()
})

missing_values['total_missing'] = (
    missing_values['null_values'] + missing_values['zero_values']
)

print("Missing values; ")
display(missing_values)

Missing values; 


,null_values,zero_values,total_missing
latitude,1,3085,3086
longitude,0,3067,3067
temperature,54,4066,4120
humidity,206,4159,4365
pm2_5_calibrated_value,325,0,325
pm10_calibrated_value,427,3,430


In [55]:
print("First five rows; ")
display(data.head())

First five rows; 


,site_name,datetime,frequency,network,latitude,longitude,temperature,humidity,pm2_5_calibrated_value,pm10_calibrated_value,site_id,device_name
0,Salama Road,2026-01-01 00:00:00Z,hourly,airgradient,0.33354,32.56861,20.535000,88.405000,181.16,197.89,694140834973ef0013d1a9b0,ag_588c81267534
1,Salama Road,2026-01-01 01:00:00Z,hourly,airgradient,0.33354,32.56861,20.640000,89.610000,290.98,318.91,694140834973ef0013d1a9b0,ag_588c81267534
2,Salama Road,2026-01-01 02:00:00Z,hourly,airgradient,0.33354,32.56861,20.821053,89.610526,274.59,301.09,694140834973ef0013d1a9b0,ag_588c81267534
3,Salama Road,2026-01-01 03:00:00Z,hourly,airgradient,0.33354,32.56861,20.605000,90.055000,242.30,264.05,694140834973ef0013d1a9b0,ag_588c81267534
4,Salama Road,2026-01-01 04:00:00Z,hourly,airgradient,0.33354,32.56861,20.675000,88.715000,206.34,225.85,694140834973ef0013d1a9b0,ag_588c81267534


In [56]:
print("Unique values in categorical columns; ")
for column in ['frequency', 'network', 'site_name', 'device_name']:
    print("\n Column: ")
    print(data[column].unique())

Unique values in categorical columns; 

 Column: 
<StringArray>
['hourly']
Length: 1, dtype: str

 Column: 
<StringArray>
['airgradient', 'airqo']
Length: 2, dtype: str

 Column: 
<StringArray>
[                                     'Salama Road',
                                  'Mbarara Central',
           'Rakai Health Sciences Program Facility',
                      'Mbarara University, Mbarara',
                                        'Kawanda 2',
                                 'Gulu Main Market',
                               'kawempe moslem p/s',
                    'Civic Centre, Kampala Central',
                         'Nsasa, Kira Municipality',
                'Embassy of France, Lumumba avenue',
 ...
                                   'Buziga Islamic',
                         'KCCA Central division HQ',
       'Golf Course Kampala. Site 1. Training Area',
                                 'Kiteezi. Kampala',
                                  'Munyonyo Site 1',
      

In [57]:
print(f"Datetime samples: {data['datetime'].head(10).tolist()}")

Datetime samples: ['2026-01-01 00:00:00Z', '2026-01-01 01:00:00Z', '2026-01-01 02:00:00Z', '2026-01-01 03:00:00Z', '2026-01-01 04:00:00Z', '2026-01-01 05:00:00Z', '2026-01-01 06:00:00Z', '2026-01-01 07:00:00Z', '2026-01-01 08:00:00Z', '2026-01-01 09:00:00Z']


In [58]:
print("Numeric description; ")
display(data.describe().T)

Numeric description; 


,count,mean,std,min,25%,50%,75%,max
latitude,14599.0,0.265731,0.680336,-1.2726,0.00000,0.313564,0.351635,2.829097
longitude,14600.0,25.272042,13.158312,0.0000,30.25565,32.552280,32.591359,33.235100
temperature,14546.0,17.295417,11.326213,0.0000,0.00000,21.600000,25.508334,44.510000
humidity,14394.0,56.100501,37.608918,0.0000,0.00000,72.720834,86.264464,99.000000
pm2_5_calibrated_value,14275.0,32.706103,30.364417,4.2900,13.25500,20.840000,44.975000,998.400000
pm10_calibrated_value,14173.0,45.079282,45.226573,0.0000,14.56000,33.890000,59.840000,802.100000


In [59]:
datetime_parsed = pd.to_datetime(['datetime'], errors='coerce')

print("Invalid datetime values; ")
print(datetime_parsed.isna().sum())

print("\n Datetime range; ")
print(f"Earliest: {datetime_parsed.min()}")
print(f"Latest: {datetime_parsed.max()}")

Invalid datetime values; 
1

 Datetime range; 
Earliest: NaT
Latest: NaT


In [60]:
print("Duplicate rows; ")
print(data.duplicated().sum())

Duplicate rows; 
0


In [61]:
print("Rows with missing pm2.5 target; ")

pm2_5 = ['pm2_5_calibrated_value']
missing_pm2_5 = pd.DataFrame({
    'null_values' : data[pm2_5].isna().sum(),
    'zero_values' : (data[pm2_5]==0).sum()
})

missing_pm2_5['total_missing'] = (
    missing_pm2_5['null_values'] + missing_pm2_5['zero_values']
)

display(missing_pm2_5)

Rows with missing pm2.5 target; 


,null_values,zero_values,total_missing
pm2_5_calibrated_value,325,0,325


In [62]:
clean_data = data.copy()

In [63]:
clean_data[measurement_columns] = (clean_data[measurement_columns].replace(0, pd.NA))
clean_data['datetime'] = pd.to_datetime(clean_data['datetime'], format='ISO8601', errors='coerce', utc=True)

In [64]:
print(f"Data shape: {clean_data.shape}")

Data shape: (14600, 12)


In [65]:
print("Invalid datetime values: ")
print(clean_data['datetime'].isna().sum())

print("\n Datetime range; ")
print(f"Earliest: {clean_data['datetime'].min()}")
print(f"Latest: {clean_data['datetime'].max()}")

Invalid datetime values: 
0

 Datetime range; 
Earliest: 2026-01-01 00:00:00+00:00
Latest: 2026-06-03 03:00:00+00:00


In [66]:
print("Missing values after cleaning zeros: ")
print(clean_data.isna().sum())

Missing values after cleaning zeros: 
site_name                    0
datetime                     0
frequency                    0
network                      0
latitude                  3086
longitude                 3067
temperature               4120
humidity                  4365
pm2_5_calibrated_value     325
pm10_calibrated_value      430
site_id                   7508
device_name                  0
dtype: int64


### Sensor Identifiers

- Uniquely identify sensors relative to their locations, to see how many unique sensor each location has

In [67]:
print("Sensor-level coverage using device_name; ")
sensor_coverage = (clean_data.groupby('device_name', dropna=False).agg(first_timestamp=('datetime', 'min'), last_timestamp=('datetime', 'max'), records=('datetime', 'count'), unique_sites=('site_name', 'nunique'), unique_site_ids=('site_id', 'nunique')).sort_values('records'))

display(sensor_coverage.head(10))

Sensor-level coverage using device_name; 


,first_timestamp,last_timestamp,records,unique_sites,unique_site_ids
device_name,,,,,
ag_588c81266dc8,2026-04-01 05:00:00+00:00,2026-06-02 11:00:00+00:00,20,1,0
airqo_g5367,2026-01-01 07:00:00+00:00,2026-01-02 15:00:00+00:00,21,1,1
airqo_g5339,2026-01-01 00:00:00+00:00,2026-06-02 00:00:00+00:00,23,1,1
aq_04,2026-06-01 06:00:00+00:00,2026-06-02 18:00:00+00:00,27,1,1
aq_48,2026-04-01 16:00:00+00:00,2026-04-02 21:00:00+00:00,30,1,0
airqo_g5564,2026-01-01 00:00:00+00:00,2026-06-02 11:00:00+00:00,33,1,0
airqo_g5204,2026-01-01 07:00:00+00:00,2026-06-02 11:00:00+00:00,34,1,1
airqo_g5305,2026-01-01 07:00:00+00:00,2026-01-02 23:00:00+00:00,37,1,1
aq_27,2026-01-01 00:00:00+00:00,2026-01-02 23:00:00+00:00,39,1,0


### Sensor-Hour uniqueness and coverage

- Check if each device has less/more than one record per hour and hence summarize the amount of data available per device

In [70]:
sensor_hour_duplicates = (
    clean_data.groupby(['device_name', 'datetime'], dropna=False).size().reset_index(name='rows_per_sensor_hour')
)

duplicate_sensor_hours = sensor_hour_duplicates[
    sensor_hour_duplicates['rows_per_sensor_hour'] > 1
]

print("Duplicate sensor-hour combinations:")
print(f"Number of duplicated sensor-hour combinations: {len(duplicate_sensor_hours)}")
print(
    f"Extra rows caused by duplicates: "
    f"{duplicate_sensor_hours['rows_per_sensor_hour'].sub(1).sum()}"
)

print("\n Records per device; ")
records_per_device = (
    clean_data.groupby('device_name', dropna=False).size().rename('records')
)

print(records_per_device.describe())

print("\n Devices with the most records: ")
print(records_per_device.sort_values(ascending=False).head(10))

print("\n Devices with the fewest records: ")
print(records_per_device.sort_values().head(10))

print("\n Sites associated with more than one device: ")
devices_per_site = (
    clean_data.groupby('site_name', dropna=False)['device_name'].nunique().sort_values(ascending=False)
)

print(devices_per_site[devices_per_site > 1 ])

Duplicate sensor-hour combinations:
Number of duplicated sensor-hour combinations: 0
Extra rows caused by duplicates: 0

 Records per device; 
count    149.000000
mean      97.986577
std       39.608385
min       20.000000
25%       58.000000
50%      105.000000
75%      137.000000
max      147.000000
Name: records, dtype: float64

 Devices with the most records: 
device_name
airqo-g5184    147
airqo_g5282    147
aq_66          147
aq_g5_37       147
aq_g519        147
airqo_g5247    146
aq_g535        145
aq_58          145
aq_g517        145
aq_47          145
Name: records, dtype: int64

 Devices with the fewest records: 
device_name
ag_588c81266dc8    20
airqo_g5367        21
airqo_g5339        23
aq_04              27
aq_48              30
airqo_g5564        33
airqo_g5204        34
airqo_g5305        37
aq_27              39
airqo_g5566        40
Name: records, dtype: int64

 Sites associated with more than one device: 
site_name
Makerere University Weather Station 94    3
Civic 

- No duplicate sensor-hour combinations, so, device-name and datetime can be used to identify an individual observation
- Sensor coverage is highly uneven, some having up to 147 records and others having only 20 records
- Multiple devices can belong to one site (hence site_name can't be a sensor identifier)

### Time gaps between consecutive observations for each sensor

In [72]:
sorted_data = clean_data.sort_values(
    ['device_name', 'datetime']
).copy()

sorted_data['time_since_previous_reading'] = (
    sorted_data.groupby('device_name')['datetime'].diff()
)

gap_summary = (
    sorted_data.dropna(subset=['time_since_previous_reading'])['time_since_previous_reading'].value_counts().sort_index()
)

print("Most common time gaps between consecutive readings; ")
print(gap_summary.head(20))

print("\n Time-gap description: ")
print(
    sorted_data['time_since_previous_reading'].dropna().describe()
)

print("\n Number of gaps longer than one hour; ")
long_gaps = sorted_data[
    sorted_data['time_since_previous_reading'] > pd.Timedelta(hours=1)
]

print(len(long_gaps))

print("\n Largest observed gaps: ")
display(
    long_gaps[
        ['device_name', 'datetime', 'time_since_previous_reading']
    ].sort_values('time_since_previous_reading', ascending=False).head(20)
)

Most common time gaps between consecutive readings; 
time_since_previous_reading
0 days 01:00:00    13668
0 days 02:00:00      236
0 days 03:00:00      125
0 days 04:00:00       43
0 days 05:00:00       44
0 days 06:00:00       25
0 days 07:00:00       19
0 days 08:00:00       14
0 days 09:00:00        7
0 days 10:00:00        5
0 days 11:00:00        1
0 days 12:00:00        4
0 days 13:00:00        7
0 days 14:00:00        3
0 days 15:00:00        3
0 days 16:00:00        3
0 days 17:00:00        6
0 days 18:00:00        5
0 days 19:00:00        2
0 days 20:00:00        4
Name: count, dtype: int64

 Time-gap description: 
count                     14451
mean     1 days 04:38:53.734689
std      9 days 08:28:41.013483
min             0 days 01:00:00
25%             0 days 01:00:00
50%             0 days 01:00:00
75%             0 days 01:00:00
max           150 days 19:00:00
Name: time_since_previous_reading, dtype: object

 Number of gaps longer than one hour; 
783

 Largest observed 

,device_name,datetime,time_since_previous_reading
12732,aq_g4_99,2026-06-02 19:00:00+00:00,150 days 19:00:00
5347,airqo-g5186,2026-04-01 05:00:00+00:00,89 days 20:00:00
8140,aq_g513,2026-04-01 00:00:00+00:00,89 days 15:00:00
6004,airqo_g5342,2026-04-01 06:00:00+00:00,89 days 12:00:00
5429,airqo_g5206,2026-04-02 09:00:00+00:00,89 days 09:00:00
9497,aq_g5_86,2026-04-02 06:00:00+00:00,89 days 08:00:00
6370,airqo_g5375,2026-04-01 00:00:00+00:00,88 days 22:00:00
6919,airqo_g5564,2026-04-01 04:00:00+00:00,88 days 18:00:00
5416,airqo_g5204,2026-04-01 06:00:00+00:00,88 days 17:00:00
8416,aq_g522,2026-04-01 15:00:00+00:00,88 days 15:00:00
